File: bees.py
import random

def log(s):
    # pass
    print s

def minimize(f):
    """
    Return a positive number that varies inversely with the function given.
    """
    def fitness(x):
        fx = f(x)
        if fx >= 0:
            return 1.0 / (fx + 1)
        else:
            return 1.0 + abs(fx)
    return fitness


class Vector:
    size = 10
    
    def __init__(self, v):
        self.v = v
        self.clamp()
        
    @staticmethod
    def new():
        v = []
        for i in xrange(Vector.size):
            v.append(random.random() * 10)
        return Vector(v)
    
    @minimize
    def fitness(self):
        mean = sum(self.v) / float(len(self.v))
        total_squared_error = 0
        for elt in self.v:
            total_squared_error += (elt - mean) ** 2
        return total_squared_error / len(self.v)

    def clamp(self):
        self.v = [max(0, min(10, elt)) for elt in self.v]

    def __iter__(self):
        return iter(self.v)
        
    @staticmethod
    def mutate(vectors, n, alpha):
        """
        contruct a mutant v' from v_n as follows:
        for each element i in v_n:
        select a random v_m such that v_m != v_n
            v'[i] = v_n[i] + phi * (v_n[i] - v_m[i])
            (where phi is a random number between -alpha and alpha)
        In other words, tweak each value in v_n either towards or away from v_m.
        """
        def phi():
            return random.random() * 2 * alpha - alpha
        vector = vectors[n]
        mutant_vector = []
        for i, elt in enumerate(vector):
            m = random.randrange(len(vectors) - 1)
            if m >= n: m += 1
            diff = elt - vectors[m].v[i]
            mutant_vector.append(elt + phi() * diff)
        return Vector(mutant_vector)

    def __str__(self):
        return '[ ' + ', '.join(("%5.2f" % e for e in self.v)) + ' ]'

    
class Solution:
    max_attempts = 10
    
    def __init__(self, vector, fitness, attempts):
        self.vector = vector
        self.fitness = fitness
        self.attempts = attempts

    @staticmethod
    def from_vector(vector):
        fitness = vector.fitness()
        attempts = Solution.max_attempts
        return Solution(vector, fitness, attempts)

    @staticmethod
    def new():
        vector = Vector.new()
        fitness = vector.fitness()
        attempts = Solution.max_attempts
        return Solution(vector, fitness, attempts)

    def __str__(self):
        return "%8.3f %3d %s" % (self.fitness, self.attempts, str(self.vector))
    
class Hive:
    def __init__(self, workers=10, observers=10, alpha=1.0):
        self.solutions = []
        for i in xrange(workers):
            self.solutions.append(Solution.new())
        self.best = max(self.solutions, key=lambda s: s.fitness)
        self.alpha = alpha
        self.observers = observers
        
    def mutate(self, n):
        vectors = [s.vector for s in self.solutions]
        mutant_vector = Vector.mutate(vectors, n, self.alpha)
        return Solution.from_vector(mutant_vector)
        
    def work_on(self, n):
        current = self.solutions[n]
        mutant = self.mutate(n)
        if mutant.fitness > current.fitness:
            self.solutions[n] = mutant
            bang = '!'
            if mutant.fitness > self.best.fitness:
                self.best = mutant
                bang = '^'
        else:
            current.attempts -= 1
            bang = ' '
        log("work on %3d %s (%2d)" % (n, bang, self.solutions[n].attempts))
            
    def worker_phase(self):
        for n in xrange(len(self.solutions)):
            self.work_on(n)
            
    def observer_phase(self):
        fitnesses = [s.fitness for s in self.solutions]
        total_fitness = sum(fitnesses)
        for o in xrange(self.observers):
            i = 0
            r = random.random() * total_fitness
            while fitnesses[i] < r:
                r -= fitnesses[i]
                i += 1
            self.work_on(i)
                
    def scout_phase(self):
        for i, solution in enumerate(self.solutions):
            if solution.attempts <= 0:
                log("reset %2d" % i)
                self.solutions[i] = Solution.new()
        
    def __str__(self):
        lines = []
        for i, s in enumerate(self.solutions):
            lines.append("%4d  %s" % (i, str(s)))
        lines.append("BEST  " + str(self.best))
        return '\n'.join(lines)
        
    def iteration(self):
        self.worker_phase()
        self.observer_phase()
        self.scout_phase()
        log(str(self))

    def run(self, n):
        for i in xrange(n):
            self.iteration()
        return self.best

if __name__ == '__main__':
    hive = Hive()
    hive.run(1000)

File: bees.ml
let log_on = ref true

let set_logging = (:=) log_on

module type GENETIC = sig
  type t
  val make         : 'a -> t
  val fitness      : t -> float
  val mutate       : t array -> int -> t
  val make_vectors : int -> t array
  val print        : t -> unit
end

module type ABC = functor (G : GENETIC) -> sig
  type t
  val create : int -> int -> t
  val iterate : t -> int -> G.t * float
end

module G : GENETIC = struct
  type t = float array

  let make _ = Array.init 10 (fun _ -> Random.float 10.0)

  let clamp x = min 10.0 (max (0.0) x)

  let sum_with f = Array.fold_left (fun total x -> total +. f x) 0.0

  let mean_squared_error t =
    let sum = Array.fold_left (+.) 0.0 t in
    let mean = sum /. float (Array.length t) in
    let deviation x = (x -. mean) ** 2.0 in
    let mean_sum = sum_with deviation t in
    mean_sum /. float (Array.length t)

  let minimize objective_function t =
    let objective_value = objective_function t in
    if objective_value >= 0.0
    then 1.0 /. (objective_value +. 1.0)
    else 1.0 +. abs_float objective_value

  let alpha = 2.0

  let fitness = minimize mean_squared_error

  let mutate array i =
    Array.mapi
      (fun index value ->
        let r = Random.int (Array.length array - 1) in
        let i' = if r < i then r else r + 1 in (* rth non-i index *)
        let dx = abs_float (value -. array.(i').(index)) in
        let phi = Random.float (2.0 *. alpha) -. alpha in
        clamp (value +. phi *. dx))
      array.(i)

  let make_vectors n = Array.init n make

  let print = Array.iter (Printf.printf " %5.2f")

end

module Make : ABC = functor (G : GENETIC) -> struct
  type t = {
    solutions    : G.t array;
    fitnesses    : float array;
    attempts     : int array;
    mutable best : G.t * float;
    observers    : int;
  }

  let max_attempts = 10

  let create workers observers =
    let solutions = Array.init workers G.make in
    let fitnesses = Array.init workers (fun i -> G.fitness solutions.(i)) in
    let attempts = Array.create workers max_attempts in
    let best = ref (solutions.(0), fitnesses.(0)) in
    for i = 1 to Array.length solutions - 1 do
      if fitnesses.(i) > snd !best then best := solutions.(i), fitnesses.(i)
    done;
    {
      solutions;
      fitnesses;
      attempts;
      best = !best;
      observers;
    }

  let print t =
    if !log_on then begin
      Array.iteri
        (fun i row ->
          Printf.printf "%3d %3d  | " i t.attempts.(i);
          G.print row;
          Printf.printf "  | %6.3f\n" (t.fitnesses.(i)))
        t.solutions;
      Printf.printf "   BEST: [ ";
      G.print (fst t.best);
      Printf.printf "  ]%7.3f\n" (snd t.best)
    end

  let work t i =
    let vector = t.solutions.(i) in
    let fitness = t.fitnesses.(i) in
    let vector' = G.mutate t.solutions i in
    let fitness' = G.fitness vector' in
    if fitness' > fitness then begin
      t.solutions.(i) <- vector';
      t.fitnesses.(i) <- fitness';
      t.attempts.(i) <- max_attempts;
      if t.fitnesses.(i) > snd t.best then
        t.best <- vector', fitness';
    end else begin
      t.attempts.(i) <- t.attempts.(i) - 1;
    end;
    if !log_on then
      Printf.printf "work %d %s (%2d)\n"
        i (if fitness' > fitness then "!" else " ") t.attempts.(i)

  let worker_phase t =
    Array.iteri
      (fun i solution -> work t i)
      t.solutions

  let observer_phase t =
    let sum = Array.fold_left (+.) 0.0 t.fitnesses in
    let rec select i r' =
      if r' < t.fitnesses.(i)
      then i
      else select (i + 1) (r' -. t.fitnesses.(i))
    in
    for o = 0 to t.observers - 1
    do
      let i = select 0 (Random.float sum) in
      work t i;
    done

  let scout_phase t =
    Array.iteri
      (fun i remaining ->
        if remaining <= 0 then begin
          if !log_on then Printf.printf "reset %d\n" i;
          t.solutions.(i) <- G.make ();
          t.fitnesses.(i) <- G.fitness t.solutions.(i);
          t.attempts.(i) <- max_attempts
        end)
      t.attempts

  let iteration t =
      worker_phase t;
      observer_phase t;
      scout_phase t;
      print t

  let rec iterate t n =
    if n = 0 then
      t.best
    else begin
      iteration t;
      iterate t (n - 1)
    end

end

module Example = Make(G)

let _ = Random.self_init ();;
let test = Example.create 10 10;;

let vec, fit = Example.iterate test 10000;;
G.print vec;
print_float fit;
print_newline ()
